### DISCUSSION TITLE
Can I have an alphabetically sorted outline?

### DISCUSSION THREAD

**Discussion by vojtechkral** [typst/typst#3431] [Q&A] [answered]
I'm investigating potential switch to Typst from LaTeX. One of the requirements is alphabetically sorted outlines/listings/indices.

Is it possible to have the output of `#outline(...)` alphabetically sorted?

**Comment by Julian-Wassmann** [comment, marked as answer]
Hi,

It is currently not possible to directly sort the entries of `outline`, like in `bibliography`. But it is possible to create a custom outline with sorted entries. This requires to reimplement the `outline.entry` element.

```typst
#let sortedOutline = () => {
  heading([Contents], outlined: false, numbering: none)
  locate(loc => {
    let entries = query(heading.where(outlined: true), loc)
    // sort entries by given key function
    let sortedEntries = entries.sorted(key: entry => {
      if entry.body.has("text") {
        entry.body.text
      } else {
        ""
      }
    })
    sortedEntries.map(entry => {
      // reimplement default outline.entry
      if entry.numbering != none {
        numbering(entry.numbering, ..counter(heading).at(entry.location()))
      }
      [ ]
      entry.body
      [ ]
      box(width: 1fr, repeat[.])
      [ ]
      [#entry.location().page()]
    }).join([ \ ])
  })
}
```

You can add additional styling or change the sorting function as you like.

If you need a sorted glossary, you can for example use the package _glossarium_ ([0.2.6](https://github.com/typst/packages/tree/main/packages/preview/glossarium/0.2.6)).

Reproduction URL: https://typst.app/project/rCAMUwkjfrxSqEzMfi6yws

**Reply by vojtechkral** [reply]
Hi. This works indeed, thank you!

I wonder if there'd be value in having support for this in the built-in outline fn, but if not, that's alright to as long as it can be solved in another.

**Reply by vojtechkral** [reply]
@Julian-Wassmann Sorry for necroposting but could I trouble you to explain why you used a lambda here rather than a regular function? (I'm going through my Typst experiments and noticed I don't quite understand this one.)

**Reply by Julian-Wassmann** [reply]
Hi! I don't think I understood what you mean with a lambda function. Can you explain this further?
Do you mean the function inside `locate`? This was necessary for Typst <= 0.10. For Typst >= 0.11 you would use the new context system for `query`. I will update my answer soon, I'm currently on the move.
Or do you mean the key function for `entries.sorted()`?

**Reply by vojtechkral** [reply]
@Julian-Wassmann I meant using `#let sortedOutline = () => { ... }` rather than `#let sortedOutline() = { .. }` :pray:
